### DISCUSSION TITLE
Refactoring the Laboratory code to a more modern codebase

### DISCUSSION THREAD

**Discussion by shanzzam** [stellar/laboratory#584] [Ideas]
As many of you know, the Laboratory is an integral part of the Stellar development ecosystem. Many people use it to experiment with new Protocol features, build transactions to sign and send to a testnet, or expand XDR into human-readable format. What you may not know is that the Lab was created in 2015 with a codebase that has remained in 2015. 

While functional, the Lab code uses outdated JS conventions/patterns and does not conform to any of our frontend style conventions. This creates some issues:

- Time consuming to ramp up on the codebase with the current structure.
- Internal Engineers change context to work on Lab as the code does not follow the same pattern as our other projects. 
- Unable to use modern JS patterns.
- No end-to-end testing
- All the above makes it difficult to add new features or redesign the UI

The SDF team maintaining the Lab codebase and features proposes a project to update the codebase. This work is not a light endeavor but luckily can be done over time and exist side-by-side with the existing codebase; this is why we're calling it a 'refactor' rather than a complete overhaul and revamp. As a side effect, we won't be taking on any new large feature set or updates during this time. The work is detailed over the following issues and broken out into 3 distinct sets of work. 

Phase 1: Setting the groundwork https://github.com/stellar/laboratory/issues/554
Phase 2: Getting code parity https://github.com/stellar/laboratory/issues/557
Phase 3: Updating the UI https://github.com/stellar/laboratory/issues/567 

We're almost compete with Phase 1 and will begin work on Phase 2 soon. This effort will be done along with other projects in our pipeline and won't have 100% of our attention. We welcome input from the community and look forward to your feedback.

**Comment by shanzzam** [comment]
Update: Phase 1 was completed a couple of months ago and we've been delayed on Phase 2. Some of the work has been started on Phase 2 and can be viewed in [this Github issue ](https://github.com/stellar/laboratory/issues/557)

We'll work on Phase 2 in the beginning of the year but don't have a solid roadmap when this work can be completed. We are open to PRs for this work from the community!

**Comment by janewang** [comment]
Closed per new Lab release https://lab.stellar.org/
